src/MongoDB.AspNetCore.OData/MongoEnableQueryAttribute.cs

```diff
@@ -0,0 +1,141 @@
+            AllowedQueryOptions.Select |
+            AllowedQueryOptions.OrderBy |
+            AllowedQueryOptions.Top |
+            AllowedQueryOptions.Skip |
+            AllowedQueryOptions.Count |
+            AllowedQueryOptions.Compute;
+    }
+
+    public override void ValidateQuery(HttpRequest request, ODataQueryOptions queryOptions)
+    {
+        base.ValidateQuery(request, queryOptions);
+
+        // Store this into the http context in case we will need to enforce projection
+        request.HttpContext.Items[nameof(ODataQueryOptions)] = queryOptions;
+    }
+
+    public override void OnActionExecuted(ActionExecutedContext actionExecutedContext)
+    {
+        base.OnActionExecuted(actionExecutedContext);
+
+        var response = actionExecutedContext.HttpContext.Response;
+        // Check whether the response is set and successful
+        if (response != null && response.IsSuccessStatusCode() && actionExecutedContext.Result is ObjectResult responseContent)
+        {
+            // If response is IMongoQueryable it means default implementation bypasses all transformations,
+            // we want force apply projection to limit MQL request to struct properties only
+            if (responseContent.Value is IMongoQueryable mongoQueryable)
+            {
+                var queryOptions = (ODataQueryOptions)response.HttpContext.Items[nameof(ODataQueryOptions)];
+                var applyProjectionMethod = s_applyProjectionMethodInfo.MakeGenericMethod(mongoQueryable.ElementType);
+                responseContent.Value = applyProjectionMethod.Invoke(null, new object[] { mongoQueryable, queryOptions });
+            }
+        }
+    }
+
+    public override IQueryable ApplyQuery(IQueryable queryable, ODataQueryOptions queryOptions)
+    {
+        if (queryable is IMongoQueryable)
+        {
+            var applyImplMethodInfo = s_applyQueryInternalMethodInfo.MakeGenericMethod(queryable.ElementType);
+            return (IQueryable)applyImplMethodInfo.Invoke(null, new object[] { queryable, queryOptions });
+        }
+
+        return base.ApplyQuery(queryable, queryOptions);
+    }
+
+    private static IQueryable ApplyQueryInternal<T>(IQueryable<T> queryable, ODataQueryOptions queryOptions)
+    {
+        var ignoreQueryOptions = AllowedQueryOptions.Select | AllowedQueryOptions.Expand;
+        queryable = (IQueryable<T>)queryOptions.ApplyTo(queryable, ignoreQueryOptions);
+
+        if (queryOptions.Count?.Value == true)
+        {
+            return queryable;
+        }
+
+        queryable = ApplyProjection(queryable, queryOptions);
+        // SelectExpand projection generated by oData library is not compatible with Mongo LINQ provider,
+        // so we have to execute the query before applying it. And then reapply Select and Expand to in-memory objects.
+        // It should be done here to turn IMongoQueryable into IEnumerable, so the detection in OnActionExecuted works as expected.
```

Comment:
The last line is not relevant anymore?

Reply:
Nope, it's still relevant. Because we still need to turn IMongoQueryable into something else, so workaround for oData default query works as expected.

Reply:
Removed as we agreed to refactor it.